src/posts/2021-02-25-Web-APIs.md

```diff
@@ -0,0 +1,168 @@
+maken van WebUSB hoeven developers maar één code base te maken voor alle platformen. Ook kunnen ze deze op afstand
+updaten waardoor het ook veilig blijft.
+
+De API is gebaseerd op de [USB 3.1 specificaties](https://www.usb.org/documents). Doormiddel van de web API beschikt
+JavaScript over alle USB functionaliteiten zoals configuraties, interfaces, endpoints en alle overdrachtstypen.
+
+#### Data uitwisseling
+
+Er zijn vier verschillende soorten van data uitwisseling die mogelijk zijn met USB. <br>
+__Control:__ De eerste soort is control. Dit protocol is gebruikt door de host om commands of query parameters te
+versturen. Binnen dit protocol worden er packets verstuurd met een lengte van 8 bytes voor een Low speed, 8-64 voor Full
+speed en 64 bytes voor High speed apparaten.
+
+__Interrupt:__ Dit protocol wordt vaak gebruikt bij apparaten die kleine hoeveelheden data versturen, bijvoorbeeld
+muizen en toetsenborden. Hierbij vraagt de host (de PC) continu data op van het externe apparaat.
+
+__Bulk:__ Voor apparaten die grote hoeveelheden data moeten versturen is er het Bulk protocol. Dit wordt vaak gebruikt
+door apparaten als printers. In dit protocol worden variabele lengte van blokken met data verstuurd of gevraagd door de
+host. Voor Full speed apparaten is de maximale lengte 64 bytes. Voor High speed apparaten is het maximum 512 bytes.
+
+__Isochronous:__ Dit protocol wordt gebruikt om real time data te streamen. Dit kan bijvoorbeeld in de vorm van
+live audio kanalen. Het protocol controleert de data niet omdat hier geen tijd voor is. De grootte van een packet kan
+maximaal 1024 bytes zijn.
+<br>
+
+Binnen het USB systeem zijn er vier soorten data packets. <br>
+__Token Packets:__ Bevat een token waardoor het systeem weet welke soort data uitwisseling er gevolgd moet worden.
+
+__Data Packets:__ Deze packets bevatten de informatie wat uitgewisseld moet worden.
+
+__Handshake Packets:__ Deze packets worden gebruikt voor het melden van errors, beantwoorden van de berichten, etc.
+
+__Start of Frame Packets:__ Deze packets worden gebruik om het begin van een nieuwe frame met data aan te geven.
+
+### Serial port
+
+Een seriële poort is een tweerichtings communicatie interface voor het versturen en ontvangen van data byte voor byte. Om
+vanaf het web toegang te krijgen tot de seriële poort van de PC moet je gebruikt maken van WebSerial API. Dit is een brug
+tussen het web en de fysieke wereld.
+
+WebSerial is onlangs gepubliceerd in Chrome 89. Om deze reden wordt het nog niet veel gebruikt en is er weinig
+documentatie over te vinden.
+
+De input en output streams worden geregeld door de Streams API. Deze API zorgt voor JavaScript objecten die het toestaan
+om programmatisch toegang te krijgen tot een data stream en die te verwerken zoals gewenst. Voor meer informatie, zie:
+[Streams API](https://developer.mozilla.org/en-US/docs/Web/API/Streams_API/Concepts).
```

Comment:
Op basis van de vorige onderwerpen, had ik hier meer informatie verwacht 😛 
Maar het is al een aardig document.

Reply:
Je zou eventueel nog een voorbeeld toepassing van WebSerial kunnen noemen? Zie bijvoorbeeld: https://github.com/svendahlstrand/web-serial-api

Reply:
Het is inderdaad al een lange blog, daarom heb ik de extra achtergrond informatie over de streams API achterwege gelaten.